app/models/conversation.rb

```diff
@@ -34,50 +34,32 @@ class Conversation < ApplicationRecord
     requested_confirmation: 1100,
   }
 
-  def api_json(include_messages = false)
-    fields = [:id, :user_id, :pickup_at, :status, :lifecycle, :from_phone,
-              :from_latitude, :from_longitude, :to_latitude, :to_longitude, :additional_passengers, :blacklisted_phone_number]
-    j = self.as_json(only: fields, methods: [:message_count])
-    if include_messages
-      j['messages'] = self.messages.map(&:api_json)
-    end
-    %w(from_address from_city to_address to_city special_requests ).each do |field|
-      j[field] = CGI::escape_html(send(field) || '')
-    end
-    last_msg = self.messages.order('created_at ASC').last
-    if last_msg
-      j['last_message_time'] = last_msg.created_at.to_i
-      j['last_message_sent_by'] = last_msg.sent_by
-      j['last_message_body'] = CGI::escape_html(last_msg.body || '')
-    end
-    j['from_phone'] = self.from_phone.phony_formatted(normalize: :US, spaces: '-')
-    j['status_updated_at'] = self.status_updated_at.to_i
-    j['name'] = CGI::escape_html(username || '')
-    j['ride'] = ride.api_json if ride
-    j['blacklisted_phone_number'] = self.blacklisted_phone.phone if self.blacklisted_phone.present?
-    j
+  def self.active_statuses
+    Conversation.statuses.keys - ['closed']
   end
 
-  # have language and name and confirmed origin and destination and confirmed time and passengers
-  def has_fields_for_ride
-    if self.user.blank? ||
-      # self.user.name.blank? ||
-      self.from_address.blank? ||
-      self.from_city.blank? ||
-      self.from_latitude.blank? ||
-      self.from_longitude.blank? ||
-      self.pickup_at.blank?
-      false
-    else
-      true
+  def self.update_ride_conversation_from_ride(ride)
+    attrs = {
+      from_address: ride.from_address,
+      from_city: ride.from_city,
+      from_state: ride.from_state,
+      from_latitude: ride.from_latitude,
+      from_longitude: ride.from_longitude,
+      to_address: ride.to_address,
+      to_city: ride.to_city,
+      to_state: ride.to_state,
+      to_latitude: ride.to_latitude,
+      to_longitude: ride.to_longitude,
+      additional_passengers: ride.additional_passengers || 0,
+      special_requests: ride.special_requests || '',
+    }
+
+    # can't change phone numbers if there are messages
+    if !ride.has_conversation_with_messages?
+      attrs[:from_phone] = ride.voter.phone_number
```

Comment:
This behavior changes with no warning to the dispatcher. That may be OK for the short-term, but it's a latent UX bug that will likely pop up at some point.

Reply:
https://github.com/john/drive.vote/pull/1085/files#diff-94d3916678683ed9ef30c06be68f0ee8R76 actually does have a warning, so I think you're covered for now. You do still have this logic in two places (model and view). At some point, it might be nice to combine them somehow, possibly with a validation.

Reply:
filed ticket: https://github.com/john/drive.vote/issues/1088